File: rainDrop/rainDrop.java
package com.me.datastructure; 
  
import java.util.Random; 
import java.util.TreeSet; 
  
public class RainDropExp { 
     static Random rdn = new Random(); 
     public static int rainDrop(){ 
         TreeSet<WetArea> set = new TreeSet<WetArea>(); 
         int i = 0; 
         while(true){ 
             oneDrop(set); 
             i++; 
             if(allWet(set)){ 
                 break; 
             } 
         } 
         return i; 
     } 
     public static void oneDrop(TreeSet<WetArea> set){ 
         double d = rdn.nextDouble(); 
         WetArea drop = new WetArea(d); 
         WetArea higher = set.ceiling(drop); 
          
         if (higher != null && drop.canMerge(higher)){ 
             drop = drop.merge(higher); 
             set.remove(higher); 
         } 
         WetArea lower = set.floor(drop); 
         if(lower != null && drop.canMerge(lower)){ 
             drop = drop.merge(lower); 
             set.remove(lower); 
         } 
         set.add(drop); 
     } 
     public static boolean allWet(TreeSet<WetArea> set){ 
         if(set.isEmpty()){ 
             return false; 
         } l
         if(set.size() == 1){ 
             return set.first().length() == 1.0; 
         } 
         return false; 
     } 
     static class WetArea implements Comparable{ 
         private double left; 
         private double right; 
         public WetArea(double centerPos){ 
             left = Math.max(centerPos - 0.005, 0); 
             right = Math.min(centerPos + 0.005, 1); 
         } 
         public WetArea(double left, double right){ 
             this.left = left; 
             this.right = right; 
         } 
         public double length(){ 
             return right-left; 
         } 
         public double centerPos(){ 
             return (right+left)/2.0; 
         } 
         public WetArea merge(WetArea another){ 
             if (this.left<=another.right && this.right >=  
another.left){ 
                 return new WetArea(Math.min(left,  
another.left), Math.max(right, another.right)); 
             } 
             throw new IllegalArgumentException(); 
         } 
         public boolean canMerge(WetArea another){ 
             return this.left<=another.right && this.right >=  
another.left; 
         } 
         /* (non-Javadoc) 
          * @see  
java.lang.Comparable#compareTo(java.lang.Object) 
          */ 
         @Override 
         public int compareTo(Object o) { 
             if (o instanceof WetArea){ 
                 WetArea o1 = (WetArea)o; 
                 double x = this.centerPos() - o1.centerPos(); 
                 if (x > 0){ 
                     return 1; 
                 }else if (x<0){ 
                     return -1; 
                 }else{ 
                     return 0; 
                 } 
                  
             } 
             throw new ClassCastException(); 
         } 
     } 
     public static void main(String[] args){ 
         int count = 10000; 
         long start = System.currentTimeMillis(); 
         int sum = 0; 
         for (int i=0; i<count; i++){ 
             sum += rainDrop(); 
         } 
         long usetime = (System.currentTimeMillis()-start)/1000; 
         System.out.println("use time=" + usetime + "s"); 
         System.out.println("avg=" + sum/(double)count); 
     } 
} 

File: rainDrop/rainDrop.py
#!/bin/python
import random
import time

class rainDrop(object):

    def __init__(self, center_pos, radius):
        super(rainDrop, self).__init__()
        self.center_pos = center_pos
        self.radius = radius
        self.left = min(center_pos - radius, 0)
        self.right = max(center_pos + radius, 1)
        self.length = 2*radius

    def merge_rain_drop(self, new_rain_drop):
        if new_rain_drop.left > self.right or \
                new_rain_drop.right < self.left:
            return False
        else:
            self.left = min(new_rain_drop.left, self.left)
            self.right = max(new_rain_drop.right, self.right)
            self.center_pos = (self.right + self.left) / 2
            self.length = (self.right - self.left)
            self.radius = self.length / 2
            return True

def binary_search(value, value_list):
    low = 0
    high = len(value_list) - 1
    while(low <= high):
        mid = (low + high)/2
        midval = value_list[mid]

        if midval < value:
            low = mid + 1
        elif midval > value:
            high = mid - 1
        else:
            return mid
    return low


class wetArea(object):

    def __init__(self, *args, **kwargs):
        super(wetArea, self).__init__()
        self.drops_pos = []
        self.drops_map = dict()

    def add_rain_drop(self, new_rain_drop):
        if self.drops_pos == []:
            self.drops_pos.append(new_rain_drop.center_pos)
            self.drops_map[new_rain_drop.center_pos] = new_rain_drop
        else:
            self.drops_pos = self.drops_pos.sort()
            new_drop_floor = self.binary_search(new_rain_drop.center_pos,
                                                self.drops_pos)
            arrive_end = True if new_drop_floor >= len(self.drops_pos) else False

            merged_drop = self.drops_map[self.drops_pos[new_drop_floor]].merge_rain_drop(new_rain_drop)
            if merged_drop:
                # pop the olds rain drop
                del self.drops_map[self.drops_pos[new_drop_floor]]
                del self.drops_pos[new_drop_floor]

                #self.drops_pos.append(merged_drop.center_pos)
                #self.drops_map[merged_drop.center_pos] = merged_drop
            else:
                #self.drops_map[new_rain_drop]
                #self.drops_pos[new_rain_drop.center_pos]
                merged_drop = new_rain_drop

            # check the ceiling one
            if not arrive_end:
                merged_drop = self.drops_map[self.drops_pos[new_drop_floor+1]].merge_rain_drop(merged_drop)
                if merged_drop:
                    del self.drops_map[self.drops_pos[new_drop_floor+1]]
                    del self.drops_pos[new_drop_floor+1]

            self.drops_pos.append(merged_drop.center_pos)
            self.drops_map[merged_drop.center_pos] = merged_drop

    def all_wet(self):
        if not self.drops_pos:
            return Flase
        if len(self.drops_pos) == 1:
            return self.drops_map[self.drops_pos[0]].length == 1

        return False

    def binary_search(self, value, value_list):
        low = 0
        high = len(value_list) - 1
        while(low <= high):
            mid = (low + high)/2
            midval = value_list[mid]

            if midval < value:
                low = mid + 1
            elif midval > value:
                high = mid - 1
            else:
                return mid
        return low


def rain_drop_expr():

    expr = wetArea()
    count = 0
    while True:
        

